### PR TITLE
add url pathmap functionality and path based routing rule

### PR DESCRIPTION
**Pull request by dilmanous**

### CODE REVIEW

**Review by mbernard — CHANGES_REQUESTED** [review]
I suggest we create another function for PathBasedRule
Add-ArmApplicationGatewayPathBasedRequestRoutingRule

**Comment by dilmanous** [comment]
> I suggest we create another function for PathBasedRule
> Add-ArmApplicationGatewayPathBasedRequestRoutingRule

the split is done and tested on my local sub

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1 [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        $HttpListenerName,
+        [string]
+        $BackendAddressPoolName,
+        [string]
+        $BackendHttpSettingsName
+    )
+
+    if (!$HttpListenerName) {
+        $HttpListenerName = $ApplicationGateway.properties.httpListeners[0].Name
+    }
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1]: Should be inside the ShouldProcess

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1 [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            }
+        }
+
+        if (!$BackendAddressPoolName) {
+            $BackendAddressPoolName = $ApplicationGateway.properties.backendAddressPools[0].Name
+        }
+    
+        if (!$BackendHttpSettingsName) {
+            $BackendHttpSettingsName = $ApplicationGateway.properties.backendHttpSettingsCollection[0].Name
+        }
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1]: Move it right after should process

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1 [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        $RequestRoutingRule.Properties.backendAddressPool = @{
+            id = "[concat($ApplicationGatewayResourceId, '/backendAddressPools/', '$BackendAddressPoolName')]"
+        }
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayBasicRequestRoutingRule.ps1]: should be done directly when we build the PSCustomObject

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayPathBasedRequestRoutingRule.ps1 [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        $Name = "default",
+        [string]
+        $HttpListenerName,
+        [string]
+        $UrlPathMapName = 'default'
+    )
+
+    if (!$HttpListenerName) {
+        $HttpListenerName = $ApplicationGateway.properties.httpListeners[0].Name
+    }
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayPathBasedRequestRoutingRule.ps1]: move after the ShouldProcess

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayPathBasedRequestRoutingRule.ps1 [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            name       = $Name
+            properties = @{
+                ruleType     = 'PathBasedRouting'
+                httpListener = @{
+                    id = "[concat($ApplicationGatewayResourceId, '/httpListeners/', '$HttpListenerName')]"
+                }
+            }
+        }
+
+        $RequestRoutingRule.Properties.urlPathMap = @{
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayPathBasedRequestRoutingRule.ps1]: move it where we build the PSCustomObject

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1 [hunk 6]
```diff
@@ -9,49 +9,55 @@ function Add-ArmApplicationGatewayRequestRoutingRule {
...
+        $BackendHttpSettingsName,
+        [PSCustomObject]
+        $PathRulesConfigs
     )
 
-    if (!$HttpListenerName) {
-        $HttpListenerName = $ApplicationGateway.properties.httpListeners[0].Name
-    }
-
     if (!$BackendAddressPoolName) {
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1]: Move everything inside the shouldProcess

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1 [hunk 7]
```diff
@@ -1,4 +1,4 @@
-function Add-ArmApplicationGatewayRequestRoutingRule {
+function Add-ArmApplicationGatewayUrlPathMap {
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1]: I don't really like how we structure that function. Let's discuss together

**Review by mbernard — CHANGES_REQUESTED** [review]

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1 (hunk 7 above)
Comment (dilmanous) [reply src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlPathMap.ps1]: added new function to add a path to a map, let me know if anything can be improved

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1 [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        $BackendAddressPoolName,
+        [Parameter(Mandatory)]
+        [string]
+        $BackendHttpSettingsName,
+        [Parameter(Mandatory)]
+        [string[]]
+        $Paths,
+        [string]
+        [Parameter(Mandatory)]
+        $MapName
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1]: UrlMapName

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1 [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        [ValidatePattern('^(\[.*\]|[a-zA-Z0-9-]*)$')]
+        [string]
+        [Parameter(Mandatory)]
+        $Name,
+        [Parameter(Mandatory)]
+        [string]
+        $BackendAddressPoolName,
+        [Parameter(Mandatory)]
+        [string]
+        $BackendHttpSettingsName,
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1]: Not sure they are mandatory

**Review by mbernard — CHANGES_REQUESTED** [review]

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1 (hunk 9 above)
Comment (dilmanous) [reply src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1]: I will try deploying without them

Comment (dilmanous) [reply src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1]: It didn't work so I had to add the default behavior to set backendpool and httpsettings

File: src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1 [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        [ValidatePattern('^(\[.*\]|[a-zA-Z0-9-]*)$')]
+        [string]
+        [Parameter(Mandatory)]
+        $Name,
+        [Parameter(Mandatory)]
+        [string[]]
+        $Paths,
+        [string]
+        [Parameter(Mandatory)]
+        $UrlMapName,
```
Comment (mbernard) [inline src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Network/ApplicationGateway/Add-ArmApplicationGatewayUrlMapPath.ps1]: not mandatory and = "default"

**Review by mbernard — CHANGES_REQUESTED** [review]

**Review by mbernard — APPROVED** [review]

**Pull request merged**